modules/rapid/src/histogram.cpp

```diff
@@ -0,0 +1,226 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+
+namespace cv
+{
+namespace rapid
+{
+static void compute1DCanny(const cv::Mat& src, cv::Mat& dst, uchar threshold)
+{
+    compute1DSobel(src, dst);
+
+    // step2: compute 1D non-maximum suppression + threshold
+    for (int i = 0; i < dst.rows; i++)
+    {
+        for (int j = 1; j < dst.cols - 1; j++)
+        {
+            if (dst.at<uchar>(i, j) <= dst.at<uchar>(i, j - 1) || dst.at<uchar>(i, j) <= dst.at<uchar>(i, j + 1))
+                dst.at<uchar>(i, j) = 0;
+
+            // threshold
+            if(dst.at<uchar>(i, j) < threshold)
+                dst.at<uchar>(i, j) = 0;
+        }
+    }
+}
+
+static void calcHueSatHist(const Mat_<Vec3b>& hsv, Mat_<float>& hist)
+{
+    for (int i = 0; i < hsv.rows; i++)
+    {
+        for (int j = 0; j < hsv.cols; j++)
+        {
+            const Vec3b& c = hsv(i, j);
+            // thresholds as in sec. 4.1
+            if (c[1] > 25 && c[2] > 50)
+            {
+                hist(c[0] * hist.rows / 256, c[1] * hist.cols / 256)++;
+            }
+        }
+    }
+}
+
+static float sum(Mat_<float>& hist)
+{
+    CV_DbgAssert(hist.isContinuous());
+    float ret = 0;
+    int N = int(hist.total());
+    const float* ptr = hist.ptr<float>();
+    for (int i = 0; i < N; i++)
+        ret += ptr[i];
+    return ret;
+}
+
+static double bhattacharyya_coeff(const Mat& a, const Mat& b)
+{
+    CV_DbgAssert(a.isContinuous() && b.isContinuous());
+    int N = int(a.total());
+    double ret = 0;
+    const float* aptr = a.ptr<float>();
+    const float* bptr = b.ptr<float>();
+    for (int i = 0; i < N; i++)
+        ret += std::sqrt(aptr[i] * bptr[i]);
+    return ret;
+}
+
+static void findCorrespondenciesOLS(const cv::Mat_<float>& scores, cv::Mat_<int>& cols)
+{
+    cols.resize(scores.rows);
+    for (int i = 0; i < scores.rows; i++)
+    {
+        int pos = -1;
+        for (int j = scores.cols - 1; j >= 0; j--)
+        {
+            if (scores(i, j) >= 0.35)
```

Comment:
I liked that you referenced where your magic numbers come from. you should do that here, too.